### PR TITLE
Ternary operator should not be nested

### PR DESCRIPTION
**Pull request by pstreef**
draft for: https://github.com/openrewrite/rewrite-static-analysis/issues/51

Recipe to replace nested ternary statements.

### CODE REVIEW

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:40]: Would be nice to override `getTags()` here to show [RSPEC-3358](https://rules.sonarsource.com/java/RSPEC-3358).

**Comment by timtebeek** [comment]
Thanks for starting yet another one! Fun to see how you're progressing. :)

It's getting a bit late for an in depth review, so I'll keep it at some general comments for now that might help you find similar recipes to copy patterns from.

I noticed you're having some excess braces / blocks inserted; while ideally these are not inserted, sometimes you can simplify your logic by inserting them unconditionally, and then cleaning up after yourself by explicitly invoking `new RemoveUnneededBlock().getVisitor().visit...(...)`. We have a few such cases for unnecessary parenthesis, but this might suit you here as well. Mostly as a way to get the tests going first, and then clean things up after.

Good seeing the tests you have already. I think it's fine to start small for instance with return statements only, and expand from there. Any tests we don't cover yet you can then annotate with `@ExpectedToFail` & `@Issue("")` referencing a new backlog item for further improvements. Benefit of only looking at `return` for now is indeed that we don't need `else` blocks yet.

**Comment by pstreef** [comment]
The thought to use something to clean the braces had crossed my mind but I thought that would be considered cheating 🙈 

Tests are green now but polishing is still required. Will leave this as a draft until and revisit it later this week (or possibly later today).

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:34]: This gave me a good chuckle. At first I thought it's not as inline with our other recipe descriptions, but it turns out it's a direct copy of the description over on: https://sonarsource.atlassian.net/browse/RSPEC-3358

Now I'm wondering if we should rewrite this to be a bit shorter and more in line, and potentially not use the description one-to-one as seen on an external source. Something like:
```suggestion
        return "Nested ternary operators can be hard to read quickly. Prefer simpler constructs for improved readability.";
```

**Comment by timtebeek** [comment]
Maybe we could add a few more tests just to be sure where the choice options are not literal strings. So for instance:
1. method calls
2. or `null`,
3. or expressions `"foo" + 123`,
4. or options preceded by a comment or newline.
 
All such cases could require slightly different handling, which could negatively affect either the functioning of code, or preservation of whitespace/comments for readability.

**Comment by rpau** [comment]
Thanks @pstreef for your contribution! 

I would like to add to the @timtebeek list, a new test for lambda expressions. It might happen that we should add a new block entry before a new if statement. 

```
 s.stream().map(item -> item.startsWith("s") ? "ss".equals(item) ?"a": "b": "c").collect(Collectors.toSet());
```

should be translated to 
```
s.stream().map(item -> { 
  if (item.startsWith("s") {
    if ("ss".equals(item)) {
       return "a";
    } else {
       return "b";
    }
    return "c";
 }
}).collect(Collectors.toSet());
```

**Comment by timtebeek** [comment]
> a new test for lambda expressions. It might happen that we should add a new block entry before a new if statement.
> 
> ```
>  s.stream().map(item -> item.startsWith("s") ? "ss".equals(item) ?"a": "b": "c").collect(Collectors.toSet());
> ```

Interesting case as well! :) Might be good to add a test that verifies no changes are made in such cases yet; so not `@ExpectedToFail`, but rather guard against accidental changes that could result in incorrect code, as the use within a lambda implies a return. Unless we're able to cover this case easily as well of course.

**Comment by pstreef** [comment]
Thanks for these notes/additions. I will take them into account.

@rpau I'm doubting if that is really a rewrite everyone would be happy with. Some people prefer to keep their streams brace-less (which is also why you might see ternaries like these pop up) and in stead of adding this in line would expect a method call.

So this would be an opinionated approach, which I guess openrewrite is full of, but I'm still doubting which way would be "best"

**Comment by rpau** [comment]
@pstreef  the problem extracting a method is that we can't infer a meaningful method name, and I agree that the strategy might be opinionated.

IMO, Skipping these cases for lambda expressions might create the impression that the recipe is not fully working. As a developer, I would like at least to identify those cases and then adapt the code.. extracting the method for instance, but if no changes are made, it is hard to identify if the code is satisfying the rule or not. 

We can add a boolean parameter to support those cases or not, but as a developer, I believe that we want to identify it to create trust in our recipes.

**Comment by pstreef** [comment]
@timtebeek the build here keeps failing, probably because it depends on stuff that is not yet merged. Since the dependant change seems to be "coming soon" perhaps it's best to just keep this open until that is rounded up?

**Comment by pstreef** [comment]
@timtebeek the build here keeps failing, probably because it depends on stuff that is not yet merged. Since the dependant change seems to be "coming soon" perhaps it's best to just keep this open until that is rounded up?

**Comment by pstreef** [comment]
@timtebeek the build here keeps failing, probably because it depends on stuff that is not yet merged. Since the dependant change seems to be "coming soon" perhaps it's best to just keep this open until that is rounded up?

**Comment by timtebeek** [comment]
Yes it's a bit unfortunate this past week.. hopefully resolved shortly! Right now you're dependent on whatever was the last snapshot pushed, which is unfortunate. We don't typically have such land lived branches, but then we also don't do many major releases. Here's hoping we can continue this soon.

**Comment by pstreef** [comment]
no worries, I will keep an eye out for the release then!

**Comment by knutwannheden** [comment]
Looking at the tests I think it would be cool if the recipe could turn some of the ternary operators into switch statements instead.

```java
                  return "a".equals(a) ? "a" : "b".equals(b) ? "b" : "nope";
```
could then be transformed to:
```java
                  return switch (a) {
                    case "a" -> "a";
                    case "b" -> "b";
                    default -> "nope";
                  };
```
or (without Java 14 switch expressions):
```java
                  switch (a) {
                    case "a":
                      return "a";
                    case "b":
                      return "b";
                    default:
                      return "nope";
                  };
```
This would of course only be applicable when all conditions always compare the same object to either literals or final variables and only be applicable to certain types (primitives, enums, and Strings). With Java 17 switch expressions with pattern matching these limitations could be relaxed a bit. The conditions would still have to refer to the same object, but the conditions could be different and also the types wouldn't have to be restricted.

Any thoughts on this?

**Comment by knutwannheden** [comment]
@pstreef Thanks for your feedback.

> But then again, wouldn't that also mean that any if that matches this should be rewritten to such a switch, so maybe a separate recipe is warranted?

I also thought a bit about this and this could indeed be an idea. Although the analysis of the `if` statements is probably harder than that of a nested ternary expression, since it isn't just a single expression anymore. But regardless of this, I also can't quite make up my mind where the lines should be drawn between these recipes.

> Note there is already a recipe that checks: `switch` statements should have at least 3 `case` clauses
> Which means that your example actually should not be a switch 😬

You are probably referring to `MinimumSwitchCases`. There the `default` case is also considered a case, so a switch with two cases and a default would not be rewritten by `MinimumSwitchCases`. But possibly there should also be an inverse recipe which rewrites a cascade of `if` statements to a `switch`, at which point we could again discuss leaving this recipe as-is.

IMHO it would be cool if we could extend this recipe to also support mapping to Java 14 switch expressions. We could then in a next step consider pulling this logic out into a separate visitor or even recipe. But I think it is also fine if we delay all this to a later commit.

**Comment by pstreef** [comment]
Hi @knutwannheden 

While investigating this possibility I found that with this PR deeper nested ternaries were not unwrapped. I'm now experimenting with a recursive approach which might also be extensible for the switch approach. Although I expect determining if a switch can be used is not so easy to implement.

While I agree it would be cool, we have to also be practical and consider how often would someone actually write such a (pretty badly) nested ternary and want it to be replaced with a switch.

I'm working now on getting all the comments properly placed with this recursive approach. After which I will update the PR.

**Comment by knutwannheden** [comment]
Sounds good. Let's wait with the switch part. Although personally I would probably already use it instead of two nested ternaries, as it has the big advantage of staying an expression (which is why I personally also sometimes like using nested ternaries, when switch expressions are not available 🙈).

**Comment by pstreef** [comment]
hi @knutwannheden
I find the challenge intriguing and since we're still blocked by v8 I can spend a little more time to try and figure it out.
 
I agree that in most of the test cases the nested ternary can be changed into a switch. I think the complexity comes from
1. Nesting can be in both true/false part. Maybe we can conclude that if there is any nesting in the truePart it's not a "switch candidate"
2. Nested ternaries can check different things, but if they all check the same variables equality to a thing it is a switch candidate.
3. If we include pattern matching for switch we can loosen the 2. constraint. But this is very dependant on the language level/confi. I'm not sure how to add this into consideration.

do you see any other checks I should add?

**Comment by knutwannheden** [comment]
@pstreef I agree, this is a challenging recipe, which is also what makes it fun to write.

> 1. Nesting can be in both true/false part. Maybe we can conclude that if there is any nesting in the truePart it's not a "switch candidate"

Agreed. Although if it turns out to make your recursive algorithm simpler, we could of course also turn this into a nested switch expression.

> 2\. Nested ternaries can check different things, but if they all check the same variables equality to a thing it is a switch candidate.

This is probably the difficult part. An equality check could when using `Object#equals(Object)` also be "reversed", using referential equality (for primitives and enums) or even use `Objects#equals(Object, Object)`.

> 3\. If we include pattern matching for switch we can loosen the 2. constraint. But this is very dependant on the language level/confi. I'm not sure how to add this into consideration.

This would be really cool, but adds more complexity and could also be left for a follow-up commit.

I can't think of any additional checks in addition to what you already identified.

**Comment by pstreef** [comment]
@knutwannheden it is indeed both fun and a challenge, but not as big a challenge as finding enough time for it. I'm well underway though!

image

**Comment by pstreef** [comment]
@knutwannheden I've found some more time. I have a lot of green tests and a few which we can either leave as `@ExpectedToFail` or maybe you can help (especially with the lost comment one)

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:241]:
```suggestion
                    maybeRemoveImport("java.util.Objects");
```

**Review by knutwannheden — COMMENTED** [review]
Apart from a few minor comments, I think the code looks very good and I think this will be a very useful recipe to have!

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (pstreef) [reply src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:89]: yes, good catch!

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (pstreef) [reply src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java]: done!

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:311]: Could be an `else if` AFAICT.

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:313]: In the body of this `if` we need to check the `J.Binary#getOperator()` result. I think currently we only support `==` (and not `!=`), so we should make sure that this is indeed also the case. A corresponding test would also be nice.

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:323]: In the future we could maybe remove this limitation. For example we could exclude any variables which are constants (i.e. where the type has the `final` and `static` flags) so that we can allow rewriting ternaries like: `v == CONST_1 ? "1" : v == CONST_2 ? "2" : ""`.

**Comment by knutwannheden** [comment]
@pstreef Sorry about these late review comments. I didn't look carefully enough the first time around.

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:348]:
```suggestion
                boolean isObjects = TypeUtils.isOfClassType(maybeObjects.getType(), "java.util.Objects");
```

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (pstreef) [reply src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:311]: Correct, will change!

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (pstreef) [reply src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:313]: good catch, I will extend!

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (pstreef) [reply src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:325]: The future is now

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (pstreef) [reply src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:296]: you are correct, I pushed a test + fix!

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (pstreef) [reply src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java]: done!

**Comment by knutwannheden** [comment]
I can't think of any other cases to test. Once we release and deploy it to public.moderne.io I wouldn't be surprised if we discover a few corner cases more, we didn't think of, but that is normal 😄

**Comment by knutwannheden** [comment]
> I'd love your input on this failing test: doReplaceMultiLevelTernariesWithComments
> I'm missing a comment that is part of a "before" of a JLeftPadded. if I use getFalsePart it automatically extracts the element and the comment is dropped. How should I get this out?

I can take a look at this. But I think we could also integrate this PR as is and fix this problem later.

File: src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNested.java:68]: No need to override the default value I suppose.
```suggestion
```

File: src/test/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNestedTest.java
Comment (timtebeek) [inline src/test/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNestedTest.java:426]: Intuitively I'd have expected this comment on the `if`, not on the `block` Would that be hard to change to reduce friction on any PRs this generates?

File: src/test/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNestedTest.java
Comment (timtebeek) [inline src/test/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNestedTest.java:807]: Don't we lose null safety here?

**Review by timtebeek — APPROVED** [review]
Impressive work! I've added a few last minor comments, but trust your judgement in resolving those before a merge.

As a last comment: Perhaps it'd be nice to clearly divide the tests over ~3 nested classes: ifs / simple switch / switch expressions.

File: src/test/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNestedTest.java
Comment (pstreef) [reply src/test/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNestedTest.java:807]: You are right, I did not know switch needs a nonNull arg.
An easy way around this is to skip `Objects.equals`, alternatively we can add a null check and early exit with the "default" case. What do you think?

Comment (timtebeek) [reply src/test/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNestedTest.java:807]: I'd say we can skip `Objects.equals` now and revisit when Java 21 is out, as Java 18+ allows for null in switch.

File: src/test/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNestedTest.java
Comment (pstreef) [reply src/test/java/org/openrewrite/staticanalysis/TernaryOperatorsShouldNotBeNestedTest.java]: I remember playing with this and doubting where to put it. It's not hard to move and I think I agree with you. It looks better on the if.

**Review by timtebeek — APPROVED** [review]

**Comment by pstreef** [comment]
Regarding the nesting. I made it a bit better for the already nested `SwitchExpressionSupported`. There is no "simple" switch, so I split expression, if and no change

**Comment by timtebeek** [comment]
Maybe good to also annotated one or two representative tests with `@DocumentExample`; that way they ought to show up in the docs (soon).

**Pull request merged**
